### PR TITLE
fix(config): Avoid parsing configuration files without interpolating secrets

### PR DESCRIPTION
**Pull request by jszwedko**
Closes: https://github.com/vectordotdev/vector/issues/20974

Reverts https://github.com/vectordotdev/vector/pull/17759 and solves the issue fixed by that PR by, instead of deserializing configuration twice, to set the global `log_schema`, defer fetching values from the `log_schema` until _after_ deserialization (previously some config structs would attempt to fetch during deserialization, but the `log_schema` may not have been deserialized by that point).

Attempting to pull from `log_schema` during deserialization is an easy bug to reintroduce, but I don't see an easy way to prohibit it.

I think it is easiest to review the first two commits independently.

### CODE REVIEW

File: src/sources/file.rs [hunk 1]
```diff
@@ -453,7 +448,12 @@ impl SourceConfig for FileConfig {
 
     fn outputs(&self, global_log_namespace: LogNamespace) -> Vec<SourceOutput> {
         let file_key = self.file_key.clone().path.map(LegacyKey::Overwrite);
-        let host_key = self.host_key.clone().path.map(LegacyKey::Overwrite);
+        let host_key = self
+            .host_key
+            .clone()
+            .unwrap_or(log_schema().host_key().cloned().into())
```
Comment (bruceg) [inline src/sources/file.rs:454]: This code sequence being repeated 8(?) times makes me think there should be a helper function for it.

**Review by bruceg — APPROVED** [review]
I have one suggestion for a helper function but otherwise this LGTM.

File: website/cue/reference/components/sources/base/docker_logs.cue [hunk 2]
```diff
@@ -45,7 +45,7 @@ base: components: sources: docker_logs: configuration: {
 			[global_host_key]: https://vector.dev/docs/reference/configuration/global-options/#log_schema.host_key
 			"""
 		required: false
-		type: string: default: "host"
+		type: string: {}
```
Comment (neko-dd) [inline website/cue/reference/components/sources/base/docker_logs.cue:48]: Is there no default value anymore?

Comment (jszwedko) [reply website/cue/reference/components/sources/base/docker_logs.cue:48]: It'll default to `log_schema.host_key` as mentioned in the option description. The field itself just doesn't have an explicit default.

Comment (jszwedko) [reply website/cue/reference/components/sources/base/docker_logs.cue:48]: (or rather it defaults to "null" which will cause it to fallback to the value of `log_schema.host_key`)

**Pull request merged**